apps/dashboard/src/contract-ui/tabs/tokens/components/airdrop-upload-erc20.tsx

```diff
@@ -0,0 +1,469 @@
+  );
+  const [noCsv, setNoCsv] = useState(false);
+
+  const reset = useCallback(() => {
+    setValidAirdrop([]);
+    setNoCsv(false);
+  }, []);
+
+  const onDrop = useCallback<Required<DropzoneOptions>["onDrop"]>(
+    (acceptedFiles) => {
+      setNoCsv(false);
+
+      const csv = acceptedFiles.find(
+        (f) => csvMimeTypes.includes(f.type) || f.name?.endsWith(".csv"),
+      );
+      if (!csv) {
+        console.error(
+          "No valid CSV file found, make sure you have an address column.",
+        );
+        setNoCsv(true);
+        return;
+      }
+
+      Papa.parse(csv, {
+        header: true,
+        complete: (results) => {
+          const data: ERC20AirdropAddressInput[] = (
+            results.data as ERC20AirdropAddressInput[]
+          )
+            .map(({ address, quantity }) => ({
+              address: (address || "").trim(),
+              quantity: (quantity || "1").trim(),
+            }))
+            .filter(({ address }) => address !== "");
+
+          if (!data[0]?.address) {
+            setNoCsv(true);
+            return;
+          }
+
+          setValidAirdrop(data);
+        },
+      });
+    },
+    [],
+  );
+
+  async function _checkIsAddress(item: ERC20AirdropAddressInput) {
+    const { address, ...rest } = item;
+    // Sending tokens to the zero address will result in an error
+    // if user wants to burn, they should use the Burn tab
+    if (address === ZERO_ADDRESS) {
+      return {
+        address,
+        resolvedAddress: address,
+        isValid: false,
+        ...rest,
+      };
+    }
+
+    let isValid = true;
+    let resolvedAddress = address;
+
+    try {
+      resolvedAddress = isAddress(address)
+        ? address
+        : await resolveAddress({ name: address, client: thirdwebClient });
+      isValid = !!resolvedAddress;
+    } catch {
+      isValid = false;
+    }
+
+    return {
+      address,
+      resolvedAddress,
+      isValid,
+      ...rest,
+    };
+  }
+
+  const normalizeQuery = useQueries({
+    queries: validAirdrop.map((o) => ({
+      queryKey: ["snapshot-check-isAddress", o.address],
+      queryFn: () => _checkIsAddress(o),
+    })),
+    combine: (results) => {
+      return {
+        data: results.map((result) => result.data),
+        pending: results.some((result) => result.isPending),
+      };
+    },
+  });
+
+  const processedData = useMemo(() => {
+    const seen = new Set();
+    const filteredData = normalizeQuery.data
+      .filter((o) => o !== undefined)
+      .filter((el) => {
+        const duplicate = seen.has(el.resolvedAddress);
+        seen.add(el.resolvedAddress);
+        return !duplicate;
+      });
+
+    const valid = filteredData.filter(({ isValid }) => isValid);
+    const invalid = filteredData.filter(({ isValid }) => !isValid);
+
+    // Make sure the invalid records are at the top so that users can see
+    const ordered = [...invalid, ...valid];
+    return ordered;
+  }, [normalizeQuery.data]);
```

Comment:
cant you do *all* of this inside the `combine` function and return only the final values you need?